feat(atdd): Registry Coherence Validator

## Issue Metadata

| Field | Value |
|-------|-------|
| Date | `2026-05-03` |
| Status | `INIT` |
| Type | `implementation` |
| Branch | `feat/registry-coherence-validator` |
| Archetypes | `coach` |
| Train | `0001-self-compliance-validate` |
| Wagon | `govern-lifecycle` |
| Feature | `rule-id-registry` |

---

## Scope

### In Scope

- New module `src/atdd/coach/utils/rule_id_registry.py` that walks `src/atdd/**/conventions/*.yaml`, parses every `rules:` array, returns the canonical index `{rule_id: {convention_path, severity, description, recipe?, introduced_in?}}`.
- New coach validator `test_rule_id_registry_coherence.py` that:
  - walks every validator file in `src/atdd/{coach,coder,tester,planner}/validators/`
  - extracts emitted `rule_id` values (regex match on `rule_id="..."`, `RULE_ID = "..."`, `Violation(rule_id="...")`)
  - asserts each emission ∈ registry
- Validator runs in **permissive mode** for v1 (warns + lists drift, exit 0). Strict mode flag (`--strict-coherence`) is ready but not yet wired into `atdd validate coach`. Strict-by-default lands as part of #389 after the bulk migration cleans the inventory.

### Out of Scope

- Migrating any existing validator or convention — that's #389.
- Loading rule metadata into validators at runtime — that's #388.
- Fixing the inventory drift the validator surfaces — diagnostic only in v1.

### Dependencies

- Substrate (already landed): #340 (`Violation` + `rule-id.convention.yaml`).
- Blocks: #389 (bulk migration uses the registry to validate work-in-progress).
- Blocked by: none.

This issue is independent of #388 (runtime binding) — they may proceed in parallel.

---

## Context

### Problem Statement

| Aspect | Current | Target | Issue |
|--------|---------|--------|-------|
| Validator emission | `Violation(rule_id="X-Y-001")` and `Violation(rule_id="COACH-PRGATE-0003")` exist in real validators | Every emitted `rule_id` exists in a convention's `rules:` array | Found in audit: at least 2 emitted IDs are not in any convention. Likely more once we look across the full validator surface. |
| Drift detection | None | Coach validator surfaces all drift in one place | Without it, coherence is descriptive (someone hopes it's right), not enforced. |
| Strict-mode toggle | N/A | `--strict-coherence` flag prepared; default is permissive in v1 | Strict-by-default before #389 migrates the inventory would block every PR. |

### User Impact

- Today's `Violation` system has unbacked rule_ids in the wild (at least 2 confirmed). The system is half-there: substrate exists, drift is invisible.
- After this lands, every PR that introduces a new `Violation` with an unregistered rule_id gets surfaced (in v1 as a warning, in #389-and-beyond as a CI failure).

### Root Cause

#340 shipped `_violation.py` and the rule-id grammar SPEC but never built the cross-check. The cross-check is the part that turns "we have rule_ids" into "rule_ids are coherent."

---

## Architecture

### Existing Patterns

| Pattern | Example File | Convention |
|---------|--------------|------------|
| Coach validator walking source | `src/atdd/coach/validators/test_rule_id_uniqueness.py` | (in-file SPEC) |
| Convention YAML walker | `src/atdd/coach/utils/repo.py` | (in-file) |
| AST-based regex extraction from validator files | `src/atdd/coder/validators/test_no_silent_exception_swallowing_python.py:_walk_handler_body` | logging.convention |

### Conceptual Model

| Term | Definition | Example |
|------|------------|---------|
| Registry | Aggregated `{rule_id: metadata}` from all conventions | built once per validate run |
| Emission | Validator code that creates a `Violation(rule_id=...)` or declares `RULE_ID = "..."` | extracted via regex |
| Drift | Emission's rule_id not in registry | reported with file:line + suggested action |
| Permissive mode | Drift surfaces as WARNING, exit 0 | v1 default |
| Strict mode | Drift surfaces as ERROR, exit 1 | switch flipped post-#389 |

### Before State

```python
# A validator can emit any string as rule_id; nothing checks it
violations.append(Violation(rule_id="X-Y-001", severity=4, ...))   # invented; no convention
violations.append(Violation(rule_id="COACH-PRGATE-0003", ...))     # exists nowhere
```

### After State

```bash
$ atdd validate coach
[WARN] rule_id_registry_coherence: 2 emissions reference unregistered rule_ids:
  src/atdd/coach/validators/test_pr_gate.py:54   COACH-PRGATE-0003   not in any convention
  src/atdd/coach/validators/fixtures/x_y/sample.py:12  X-Y-001  appears to be a placeholder
  Run with --strict-coherence to fail CI on these.
```

---

## Phases

### Phase 1: Registry walker

**Deliverables:**
- `src/atdd/coach/utils/rule_id_registry.py` with `build_registry() -> Dict[str, RuleMetadata]`.
- Unit tests covering: walks every `*.convention.yaml`, parses `rules:` arrays, ignores non-rule entries, surfaces malformed YAML with location.

**Files:**

| File | Change |
|------|--------|
| `src/atdd/coach/utils/rule_id_registry.py` | New |
| `src/atdd/coach/utils/tests/test_rule_id_registry.py` | New |

### Phase 2: Coherence validator

**Deliverables:**
- `test_rule_id_registry_coherence.py` walks `src/atdd/{coach,coder,tester,planner}/validators/*.py`, extracts emissions, cross-checks against registry.
- Permissive mode default; `--strict-coherence` flag for CI.

**Files:**

| File | Change |
|------|--------|
| `src/atdd/coach/validators/test_rule_id_registry_coherence.py` | New |
| `src/atdd/cli.py` | Wire `--strict-coherence` flag on `atdd validate coach` |

---

## Validation

### Gate Tests

| ID | Phase | Command | Expected | ATDD Validator | Status |
|----|-------|---------|----------|----------------|--------|
| GT-001 | design | `atdd validate coach` | PASS | `src/atdd/coach/validators/test_issue_validation.py` | TODO |
| GT-002 | design | `atdd registry update --check` | PASS | `src/atdd/coach/commands/registry.py` | TODO |
| GT-100 | RED | `pytest src/atdd/coach/utils/tests/test_rule_id_registry.py -v` | FAIL (registry walker not implemented) | the new tests | TODO |
| GT-200 | GREEN | `pytest src/atdd/coach/utils/tests/test_rule_id_registry.py -v` | PASS | the new tests | TODO |
| GT-300 | GREEN | `pytest src/atdd/coach/validators/test_rule_id_registry_coherence.py -v` | PASS in permissive mode (drift logged but doesn't fail) | the new validator | TODO |
| GT-400 | SMOKE | `atdd validate coach` against current main | runs without failure; prints WARN list of drift (expected: 2+ emissions surfaced) | runtime | TODO |
| GT-500 | SMOKE | `atdd validate coach --strict-coherence` against current main | exits 1 with drift list | runtime | TODO |
| GT-800 | completion | `atdd urn validate` | PASS | `src/atdd/coach/validators/test_urn_traceability.py` | TODO |
| GT-850 | completion | `atdd registry update --check` | PASS | `src/atdd/coach/commands/registry.py` | TODO |
| GT-900 | completion | `atdd validate` | PASS | `src/atdd/` | TODO |

### Success Criteria

- [ ] Registry walker returns the canonical `{rule_id: metadata}` index for every convention.
- [ ] Coherence validator surfaces every emission whose rule_id is not in the registry.
- [ ] Permissive mode is default; `--strict-coherence` flag exists and is documented.
- [ ] At least the 2 known emissions (`COACH-PRGATE-0003`, `X-Y-001`) appear in the WARN list when run on current main.

---

## Decisions

| # | Question | Decision | Rationale |
|---|----------|----------|-----------|
| 1 | AST-based emission extraction or regex? | Regex on source for v1 | Emissions are simple string literals (`rule_id="..."`, `RULE_ID = "..."`); regex is fast and adequate. AST parser deferred to follow-on if false-positive rate exceeds 1%. |
| 2 | Should the validator emit its own `Violation` records on drift? | No — drift is a config issue, not a runtime violation. Plain text WARNING is right. | Avoids meta-recursion (validator emitting violations about emissions). |
| 3 | Strict mode toggled by env var, CLI flag, or convention? | CLI flag `--strict-coherence` | Easiest to reason about per-PR; conventions are already overloaded. |
| 4 | Include test fixtures in the scan, or only production validators? | Skip `**/fixtures/**` and `**/tests/**` | Fixtures contain placeholder rule_ids by design. |

---

## Activity Log

### Entry 1 (2026-05-03)

**Completed:**
- Issue created (parent: #386).
- Body populated; pending 2-pass review.

**Next:**
- 2-pass review.
- PLANNED transition.
- Spawn implementation agent in parallel with #388.

---

## Artifacts

### Created

- `src/atdd/coach/utils/rule_id_registry.py`
- `src/atdd/coach/utils/tests/test_rule_id_registry.py`
- `src/atdd/coach/validators/test_rule_id_registry_coherence.py`

### Modified

- `src/atdd/cli.py` — `--strict-coherence` flag

### Deleted

- *no deletions*

---

## Release Gate

- [ ] Rebase on main: `git pull origin main --rebase`
- [ ] Bump version (feat/ → MINOR — adds CLI flag and coach validator): edit `pyproject.toml`, commit "Bump version to X.Y.Z"
- [ ] Merge PR → CI creates tag + publishes

---

## Notes

This issue is the diagnostic half of the rule-ID coherence foundation (parent: #386). Sibling #388 (runtime binding) closes the constants-vs-convention drift class. #389 (bulk migration) flips this validator to strict mode after cleaning the inventory.

# Review of issue #387

## Verdict
NEEDS-MINOR-FIXES

The architectural premise is sound (registry walker → cross-check validator, permissive default). Template compliance is clean (`atdd issue 387 --check` passes; `atdd orchestrate 387 --dry-run` shows no cycles). But the body contains several factual errors and underspecified pieces that would mislead the implementation agent. All fixes are surgical — no need to rethink the approach.

## Findings (per checklist letter)

### A. Compliance — PASS
- `atdd issue 387 --check` reports "template compliant".
- All required H2 sections present (Issue Metadata, Scope, Context, Architecture, Phases, Validation, Decisions, Activity Log, Artifacts, Release Gate, Notes).
- `### Dependencies` block is present and concrete (substrate=#340, blocks=#389, parallel-with=#388).
- No `(define ...)` / `TBD` / `(none yet)` placeholders.
- GT-001 → GT-900 rows are concrete (commands and validators named).

### B. Ambiguities — FAIL (3 issues)

**B1. Regex extraction spec is ambiguous.**
Scope says "regex match on `rule_id="..."`, `RULE_ID = "..."`, `Violation(rule_id="...")`". Real code uses constants whose names *start with* `RULE_ID` (e.g. `RULE_ID_PRGATE_GREEN`, `XSS_RULE_ID`, `RULE_DYNAMIC_TRAIN_ID`, `RULE_HARNESS_ERROR`) — none of which match a literal `RULE_ID =` prefix. The body must specify whether the regex is anchored (`^RULE_ID =`), prefix-match (`^RULE_ID[A-Z_]* =`), or suffix-match (`[A-Z_]*RULE_ID[A-Z_]* =`). Verified against the inventory:
- `RULE_ID_PRGATE_GREEN = "COACH-PRGATE-0003"` (test_pr_phase_alignment.py:76)
- `XSS_RULE_ID = "SECURITY-XSS-001"` (test_frontend_security_patterns.py:109)
- `RULE_ID = "COACH-SILENT-SWALLOW-001"` (×2 in silent-swallow validators)
- `RULE_DYNAMIC_TRAIN_ID`, `RULE_UNREGISTERED_TRAIN`, `RULE_UNREGISTERED_WAGON`, `RULE_HARNESS_ERROR`, `RULE_STUB_RENDER`, `RULE_EMPTY_RENDER`, `RULE_ALLOWLIST_MIGRATION` — none would match `^RULE_ID = `.

**B2. "Convention `rules:` array" is structurally ambiguous.**
Phase 1 deliverable says walker "parses every `rules:` array". But surveyed `rules:` blocks have at least three shapes:
- list-of-objects: `rules: [{id: "LOG-001", ...}]` (logging:63, source-layout:39, frontend:605, refactor:547, orchestration:152, etc.)
- mapping keyed by id: `rules: {COACH-SILENT-SWALLOW-001: {...}}` (logging:24)
- mapping keyed by semantic name (no `id:` at all): `rules: {worktree_per_issue: {rule, anti_pattern, ...}}` (orchestration:10, error-response:12, smoke:44, coverage:5)

The walker spec must declare which shape(s) are canonical and what to do with the others (skip? coerce? emit a config-error?). The body currently implies a single shape.

**B3. After-State example references files that do not exist.**
The example output cites:
- `src/atdd/coach/validators/test_pr_gate.py:54` — no such file. The actual validator is `src/atdd/coach/validators/test_pr_phase_alignment.py` (line 76 declares the constant; line 159 emits the Violation). Either the example is fictional or it refers to a different validator that has not been written yet.
- `src/atdd/coach/validators/fixtures/x_y/sample.py:12` — no such fixture. `X-Y-001` actually lives in `src/atdd/coach/validators/tests/test_violation.py`.

### C. Gaps — FAIL (3 issues)

**C1. Scan scope misses real emission sites.**
Phase-2 deliverable scans `src/atdd/{coach,coder,tester,planner}/validators/*.py`. Real `rule_id=` emissions also exist in:
- `src/atdd/coach/commands/babysit.py:192,286,294` (uses `Violation` directly inside a CLI command)
- `src/atdd/coder/baselines/test_ratchet.py:50`
- `src/atdd/coder/validators/route_train_wagon_analyzer.py:173,188,201` (helper module, not a `test_*.py`)
- `src/atdd/coder/validators/presentation_ratchet.py:171` (helper module, not a `test_*.py`)

If the walker only globs `validators/*.py` it will also miss helpers (`route_train_wagon_analyzer.py`, `presentation_ratchet.py`) — those are siblings to test files, not nested. Decide whether to scan `validators/**/*.py` and `commands/**/*.py` or to keep the narrow scope and document the omission.

**C2. Success Criteria contradicts Decision #4 about `X-Y-001`.**
Decision #4 says: "Skip `**/fixtures/**` and `**/tests/**`". `X-Y-001` is only present in `src/atdd/coach/validators/tests/test_violation.py` — i.e. inside the path that Decision #4 excludes. The Success Criteria asserts "At least the 2 known emissions (`COACH-PRGATE-0003`, `X-Y-001`) appear in the WARN list when run on current main." Per Decision #4 only `COACH-PRGATE-0003` should surface. Either widen the scan to include `tests/test_violation.py`, or drop `X-Y-001` from the success criteria. Recommend the latter (keep tests excluded; it's cleaner).

**C3. Validator path for own unit tests is fine but unstated as "new".**
`src/atdd/coach/utils/tests/` exists (verified via `ls`); the new file `test_rule_id_registry.py` does not — should be marked "New" in the Files table, which it already is. **PASS** for this sub-item; flagging only because the surrounding cluster has issues.

### D. Regression risk — PASS (low)

- New files only; no existing convention/validator is mutated. Files modified:
  - new `src/atdd/coach/utils/rule_id_registry.py` (verified absent)
  - new `src/atdd/coach/utils/tests/test_rule_id_registry.py` (verified absent; parent dir exists)
  - new `src/atdd/coach/validators/test_rule_id_registry_coherence.py` (verified absent)
  - additive flag in `src/atdd/cli.py` (no behavior change unless flag is passed)
- Permissive default keeps `atdd validate coach` exit code unchanged on current main. CI will not start failing because of this issue alone.
- No conflict with substrate from #340–#358:
  - `_violation.py` shape is read, not modified
  - `rule-id.convention.yaml` is read, not modified
  - No overlap with babysit allowlist, ratchet, smoke gate, or auto-phase workflow
- Existing coach validators that touch the same files: `test_rule_id_uniqueness.py` reads the same `rule-id.convention.yaml`. The new walker (`build_registry()`) should reuse or be consistent with the parsing in `test_rule_id_uniqueness.py:_load_rule_id_yaml` to avoid two slightly different YAML interpretations. Body should call this out (see proposed rewrite).
- `.atdd/config.yaml` is not modified — additive only.

### E. Coherence with today's substrate — MOSTLY PASS

- ✅ Uses `Violation` records and `rule_id` schema from #340 — correctly references `_violation.py`.
- ✅ Does not touch babysit/orchestrate UX (#343), so pane-mode multiplexer is N/A.
- ✅ Does not mutate `.atdd/manifest.yaml`, so manifest-commit helper from #344 is N/A.
- ⚠️ The new validator emits warnings in plain text rather than structured `Violation` records (Decision #2 makes this explicit and gives a sound reason — avoiding meta-recursion). Consequence: this validator does NOT itself declare a rule_id + severity. That is consistent with Decision #2 but means the new validator is an **outlier** under the post-#340 norm. Worth a one-line note in the body that this is a deliberate exception, so a future reader doesn't try to "fix" it.
- ⚠️ Permissive-vs-strict toggle is a CLI flag (Decision #3). The substrate from #345/#347 around suppression markers (`# atdd:suppress(RULE-ID)`) is a separate mechanism for suppressing per-emission, not per-validator-strictness. Body could clarify these are orthogonal so reviewers don't confuse them.

---

## Proposed body rewrite

```markdown
## Issue Metadata

| Field | Value |
|-------|-------|
| Date | `2026-05-03` |
| Status | `INIT` |
| Type | `implementation` |
| Branch | `feat/registry-coherence-validator` |
| Archetypes | `coach` |
| Train | `0001-self-compliance-validate` |
| Wagon | `govern-lifecycle` |
| Feature | `rule-id-registry` |

---

## Scope

### In Scope

- New module `src/atdd/coach/utils/rule_id_registry.py` exposing `build_registry() -> dict[str, RuleMetadata]`. The walker reads every `src/atdd/**/conventions/*.yaml`, normalizes the heterogeneous `rules:` shapes (see Architecture → Conceptual Model) into a single `{rule_id: {convention_path, severity, description, recipe?, introduced_in?}}` dict, and reuses the YAML loader from `src/atdd/coach/validators/test_rule_id_uniqueness.py` to keep parsing consistent.
- New coach validator `src/atdd/coach/validators/test_rule_id_registry_coherence.py` that:
  - walks `src/atdd/{coach,coder,tester,planner}/validators/**/*.py` AND `src/atdd/coach/commands/**/*.py` (commands also emit `Violation` — verified: `babysit.py` lines 192/286/294)
  - extracts emissions via the regex set defined below (see Decision #1)
  - cross-checks each extracted rule_id against the registry
  - emits a plain-text WARN block listing drift; exits 0 in permissive mode (default), exits 1 with `--strict-coherence`
- New CLI flag `--strict-coherence` on `atdd validate coach` in `src/atdd/cli.py`, defaulting to off.

### Out of Scope

- Migrating any existing validator or convention — that's #389.
- Loading rule metadata into validators at runtime — that's #388.
- Fixing the inventory drift the validator surfaces — diagnostic only in v1.
- Making this validator emit structured `Violation` records — Decision #2 keeps it as plain-text warnings to avoid meta-recursion.

### Dependencies

- Substrate (already landed): #340 (`Violation` + `rule-id.convention.yaml`).
- Blocks: #389 (bulk migration uses the registry).
- Blocked by: none.
- Independent of #388 (runtime binding) — may proceed in parallel.

---

## Context

### Problem Statement

| Aspect | Current | Target | Issue |
|--------|---------|--------|-------|
| Validator emission | Production validators emit rule_ids both as inline string literals (`Violation(rule_id="COACH-PRGATE-0003", ...)`) and via constants (`RULE_ID_PRGATE_GREEN = "COACH-PRGATE-0003"`); some IDs are not present in any convention's `rules:` block | Every emitted rule_id must exist in some convention `rules:` entry | Confirmed drift example: `COACH-PRGATE-0003` (declared in `test_pr_phase_alignment.py:76`, emitted line 159) is not present in any convention's `rules:` array. Likely more once the full validator+commands surface is scanned. |
| Drift detection | None | Coach validator surfaces all drift in one place | Without it, coherence is descriptive (someone hopes it's right), not enforced. |
| Strict-mode toggle | N/A | `--strict-coherence` flag prepared; default is permissive in v1 | Strict-by-default before #389 migrates the inventory would block every PR. Strict-by-default flips in #389. |

### User Impact

- Today's `Violation` system has unbacked rule_ids in the wild (at least `COACH-PRGATE-0003` confirmed). The system is half-there: substrate exists, drift is invisible.
- After this lands, every PR that introduces a new `Violation` with an unregistered rule_id gets surfaced — in v1 as a warning, in #389-and-beyond as a CI failure when the flag flips.

### Root Cause

#340 shipped `_violation.py` and the rule-id grammar SPEC but never built the cross-check. The cross-check is the part that turns "we have rule_ids" into "rule_ids are coherent."

---

## Architecture

### Existing Patterns

| Pattern | Example File | Convention |
|---------|--------------|------------|
| Coach validator walking source | `src/atdd/coach/validators/test_rule_id_uniqueness.py` | rule-id.convention.yaml |
| Convention YAML walker / loader (REUSE) | `src/atdd/coach/validators/test_rule_id_uniqueness.py:_load_rule_id_yaml` | rule-id.convention.yaml |
| Regex extraction over validator source | `src/atdd/coder/validators/test_no_silent_exception_swallowing_python.py` | logging.convention.yaml |

### Conceptual Model

| Term | Definition | Example |
|------|------------|---------|
| Registry | Aggregated `{rule_id: metadata}` from all conventions, normalized across the three observed `rules:` shapes (see below) | built once per validate run |
| `rules:` shape A (list-of-objects) | `rules: [ { id: "LOG-001", ... }, ... ]` | `coder/conventions/logging.convention.yaml:63` |
| `rules:` shape B (id-keyed mapping) | `rules: { COACH-SILENT-SWALLOW-001: { ... } }` | `coder/conventions/logging.convention.yaml:24` |
| `rules:` shape C (semantic-keyed mapping, no rule_id) | `rules: { worktree_per_issue: { rule, anti_pattern, enforced_by } }` | `coach/conventions/orchestration.convention.yaml:10` (top-level) |
| Emission | A line in production source that creates a `Violation(rule_id=...)` or declares a constant whose value is a rule_id (see Decision #1 for exact regex set) | `RULE_ID_PRGATE_GREEN = "COACH-PRGATE-0003"` |
| Drift | An emission whose extracted rule_id is not present in the registry | reported with file:line + the unmatched rule_id |
| Permissive mode | Drift surfaces as WARNING, exit 0 | v1 default |
| Strict mode | Drift surfaces as ERROR, exit 1 | enabled with `--strict-coherence`, flipped on by default in #389 |

Walker shape-handling rules:
- Shape A and B contribute rule_ids to the registry.
- Shape C is **skipped** (no rule_id field by construction). The walker may optionally log "skipped N entries in {file} (shape C — no id field)" for debuggability.
- Mixed-shape file (e.g. `logging.convention.yaml` has top-level shape-B mapping AND a nested shape-A list): both contribute; duplicates are merged with last-write-wins and a debug note.

### Before State

```python
# A validator can emit any string as rule_id; nothing checks it
violations.append(Violation(rule_id="COACH-PRGATE-0003", severity=4, ...))   # not in any rules: array
violations.append(Violation(rule_id=RULE_ID_PRGATE_GREEN, ...))               # constant whose value is also unregistered
```

### After State

```bash
$ atdd validate coach
[WARN] rule_id_registry_coherence: 1 emission references an unregistered rule_id:
  src/atdd/coach/validators/test_pr_phase_alignment.py:76   COACH-PRGATE-0003   not in any convention rules: block
  Run with --strict-coherence to fail CI on this.

$ atdd validate coach --strict-coherence
[ERROR] rule_id_registry_coherence: 1 emission references an unregistered rule_id:
  src/atdd/coach/validators/test_pr_phase_alignment.py:76   COACH-PRGATE-0003
exit 1
```

(The exact count after implementation may exceed 1 once the wider scan surface — `validators/**/*.py` plus `coach/commands/**/*.py` — is in play. Expect ≥1.)

---

## Phases

### Phase 1: Registry walker

**Deliverables:**
- `src/atdd/coach/utils/rule_id_registry.py` with `build_registry() -> dict[str, RuleMetadata]`.
- Unit tests covering: walks every `*.convention.yaml`; correctly normalizes shape A and B; skips shape C with a debug note; surfaces malformed YAML with file path + line; reuses `_load_rule_id_yaml` from `test_rule_id_uniqueness.py` for the `rule-id.convention.yaml` itself.

**Files:**

| File | Change |
|------|--------|
| `src/atdd/coach/utils/rule_id_registry.py` | New |
| `src/atdd/coach/utils/tests/test_rule_id_registry.py` | New |

### Phase 2: Coherence validator

**Deliverables:**
- `test_rule_id_registry_coherence.py` walks `src/atdd/{coach,coder,tester,planner}/validators/**/*.py` and `src/atdd/coach/commands/**/*.py`, extracts emissions via the Decision #1 regex set, cross-checks against the registry built in Phase 1.
- `**/fixtures/**` and `**/tests/**` excluded from the scan (see Decision #4).
- Permissive mode default; `--strict-coherence` flag wired through `atdd validate coach` in `src/atdd/cli.py`.

**Files:**

| File | Change |
|------|--------|
| `src/atdd/coach/validators/test_rule_id_registry_coherence.py` | New |
| `src/atdd/cli.py` | Add `--strict-coherence` option to the `validate coach` subcommand |

---

## Validation

### Gate Tests

| ID | Phase | Command | Expected | ATDD Validator | Status |
|----|-------|---------|----------|----------------|--------|
| GT-001 | design | `atdd issue 387 --check` | PASS | `src/atdd/coach/validators/test_issue_validation.py` | TODO |
| GT-002 | design | `atdd registry update --check` | PASS | `src/atdd/coach/commands/registry.py` | TODO |
| GT-100 | RED | `pytest src/atdd/coach/utils/tests/test_rule_id_registry.py -v` | FAIL (registry walker not implemented) | the new tests | TODO |
| GT-200 | GREEN | `pytest src/atdd/coach/utils/tests/test_rule_id_registry.py -v` | PASS | the new tests | TODO |
| GT-300 | GREEN | `pytest src/atdd/coach/validators/test_rule_id_registry_coherence.py -v` | PASS in permissive mode (drift logged but doesn't fail) | the new validator | TODO |
| GT-400 | SMOKE | `atdd validate coach` against current main | runs without failure; prints WARN list of drift (expected: ≥1 emission surfaced — `COACH-PRGATE-0003` is the confirmed example) | runtime | TODO |
| GT-500 | SMOKE | `atdd validate coach --strict-coherence` against current main | exits 1 with drift list | runtime | TODO |
| GT-800 | completion | `atdd urn validate` | PASS | `src/atdd/coach/validators/test_urn_traceability.py` | TODO |
| GT-850 | completion | `atdd registry update --check` | PASS | `src/atdd/coach/commands/registry.py` | TODO |
| GT-900 | completion | `atdd validate` | PASS | `src/atdd/` | TODO |

### Success Criteria

- [ ] Registry walker returns the canonical `{rule_id: metadata}` index for every convention, correctly normalizing shape A and B and skipping shape C.
- [ ] Coherence validator surfaces every emission whose rule_id is not in the registry.
- [ ] Permissive mode is default; `--strict-coherence` flag exists and is documented in `atdd validate --help`.
- [ ] At least `COACH-PRGATE-0003` appears in the WARN list when run on current main.
- [ ] Validator scans both `validators/**/*.py` and `coach/commands/**/*.py`; excludes `**/tests/**` and `**/fixtures/**`.

---

## Decisions

| # | Question | Decision | Rationale |
|---|----------|----------|-----------|
| 1 | What regex set extracts emissions? | Three patterns, all on `*.py` source: (a) `Violation\(\s*rule_id\s*=\s*"([A-Z][A-Z0-9-]+)"`, (b) `\b([A-Z_]*RULE_ID[A-Z_]*)\s*=\s*"([A-Z][A-Z0-9-]+)"` (captures both the constant name and the rule_id literal), (c) `\brule_id\s*=\s*"([A-Z][A-Z0-9-]+)"` for catch-all keyword-arg literals. AST-based extraction deferred to a follow-on if false-positive rate exceeds 1%. | Real code uses constants whose names start with `RULE_ID_*` or end with `_RULE_ID` (e.g. `RULE_ID_PRGATE_GREEN`, `XSS_RULE_ID`). Pattern (b) covers both. Patterns (a) and (c) cover inline literals. |
| 2 | Should the validator emit its own `Violation` records on drift? | No — drift is a config issue, not a runtime violation. Plain text WARN/ERROR is right. | Avoids meta-recursion (validator emitting violations about emissions). Deliberate exception to the post-#340 norm; do not "fix" it. |
| 3 | Strict mode toggled by env var, CLI flag, or convention? | CLI flag `--strict-coherence` on `atdd validate coach`. | Easiest to reason about per-PR; conventions are already overloaded. Orthogonal to per-emission `# atdd:suppress(...)` markers. |
| 4 | Include test fixtures and unit tests in the scan, or only production? | Skip `**/fixtures/**` and `**/tests/**`. | Fixtures contain placeholder rule_ids by design (e.g. `X-Y-001` in `coach/validators/tests/test_violation.py`). Production scan surface = `validators/**/*.py` + `coach/commands/**/*.py`. |
| 5 | How does the walker handle the heterogeneous `rules:` shapes? | Normalize shapes A (list-of-objects with `id:`) and B (id-keyed mapping); skip shape C (semantic-keyed mapping with no `id:` field) with a debug note; merge mixed-shape files with last-write-wins. | Surveyed conventions show all three shapes in use (logging has both A and B; orchestration top-level is shape C). Walker must commit to a normalization or it cannot build a usable registry. |

---

## Activity Log

### Entry 1 (2026-05-03)

**Completed:**
- Issue created (parent: #386).
- Body populated; pending 2-pass review.

**Next:**
- 2-pass review.
- PLANNED transition.
- Spawn implementation agent in parallel with #388.

---

## Artifacts

### Created

- `src/atdd/coach/utils/rule_id_registry.py`
- `src/atdd/coach/utils/tests/test_rule_id_registry.py`
- `src/atdd/coach/validators/test_rule_id_registry_coherence.py`

### Modified

- `src/atdd/cli.py` — `--strict-coherence` option on `atdd validate coach`

### Deleted

- *no deletions*

---

## Release Gate

- [ ] Rebase on main: `git pull origin main --rebase`
- [ ] Bump version (feat/ → MINOR — adds CLI flag and coach validator): edit `pyproject.toml`, commit "Bump version to X.Y.Z"
- [ ] Merge PR → CI creates tag + publishes

---

## Notes

This issue is the diagnostic half of the rule-ID coherence foundation (parent: #386). Sibling #388 (runtime binding) closes the constants-vs-convention drift class. #389 (bulk migration) flips this validator to strict mode after cleaning the inventory.
```

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -4,7 +4,7 @@ build-backend = "setuptools.build_meta"
 
 [project]
 name = "atdd"
-version = "1.65.8"
+version = "1.65.9"
 description = "ATDD Platform - Acceptance Test Driven Development toolkit"
 requires-python = ">=3.10"
 readme = "README.md"

--- a/src/atdd/cli.py
+++ b/src/atdd/cli.py
@@ -392,6 +392,17 @@ Phase descriptions:
             "(presentation-layer ratchet improvements over 20%%). Issue #358."
         ),
     )
+    validate_parser.add_argument(
+        "--strict-coherence",
+        action="store_true",
+        dest="strict_coherence",
+        help=(
+            "Fail (exit 1) when the rule-id registry coherence validator "
+            "finds emissions whose rule_id is not declared in any "
+            "convention rules: block. Default is permissive (drift logs "
+            "as WARN, exit 0). Active when phase=coach or all. Issue #387."
+        ),
+    )
 
     # ----- atdd inventory -----
     inventory_parser = subparsers.add_parser(
@@ -1469,6 +1480,18 @@ Phase descriptions:
         # Propagate --no-cache to GraphBuilder via env var
         if getattr(args, 'no_cache', False):
             os.environ["ATDD_NO_CACHE"] = "1"
+
+        # --strict-coherence: opt the rule-id registry coherence validator
+        # into failure mode (default is permissive WARN). Issue #387.
+        if getattr(args, 'strict_coherence', False):
+            if args.phase not in ("coach", "all"):
+                print(
+                    "Error: --strict-coherence is only supported with phase "
+                    "'coach' (or 'all'). Re-run: atdd validate coach "
+                    "--strict-coherence"
+                )
+                return 1
+            os.environ["ATDD_STRICT_COHERENCE"] = "1"
 
         coach = ATDDCoach(repo_root=repo_path)
         skip_api = getattr(args, 'skip_api', False)

--- a/src/atdd/coach/utils/rule_id_registry.py
+++ b/src/atdd/coach/utils/rule_id_registry.py
@@ -136,6 +136,11 @@ def _extract_from_rules_block(
             _logger.debug(
                 "rule_id_registry: skipped %d entries in %s (shape C — no id field)",
                 skipped, file_path,
+                extra={
+                    "skipped_count": skipped,
+                    "convention_path": str(file_path),
+                    "shape": "C",
+                },
             )
 
 
@@ -149,7 +154,11 @@ def _load_yaml(path: Path):
         with open(path) as fh:
             return yaml.safe_load(fh)
     except (OSError, yaml.YAMLError) as exc:
-        _logger.debug("rule_id_registry: skipping malformed YAML %s: %s", path, exc)
+        _logger.debug(
+            "rule_id_registry: skipping malformed YAML %s: %s",
+            path, exc,
+            extra={"convention_path": str(path), "error_type": type(exc).__name__},
+        )
         return None
 
 

--- a/src/atdd/coach/utils/rule_id_registry.py
+++ b/src/atdd/coach/utils/rule_id_registry.py
@@ -1,0 +1,194 @@
+# URN: component:govern-lifecycle:enforcement-substrate:rule_id_registry:backend:domain
+# Runtime: python
+# Purpose: Aggregate every convention's rules: block into a single {rule_id: metadata} index.
+
+"""
+Rule-ID registry walker (issue #387, Phase 1).
+
+Reads every ``src/atdd/**/conventions/*.yaml`` and normalizes the three
+observed ``rules:`` shapes (see ``src/atdd/coach/conventions/rule-id.convention.yaml``)
+into a single ``{rule_id: RuleMetadata}`` index:
+
+  Shape A  rules: [ { id: "GREEN-URN-001", ... }, ... ]              → contributes
+  Shape B  rules: { COACH-SILENT-SWALLOW-001: { id: ..., ... } }      → contributes
+  Shape C  rules: { worktree_per_issue: { rule, anti_pattern, ... } } → SKIPPED
+
+Mixed-shape files (e.g. ``logging.convention.yaml`` has both A and B) are
+merged with last-write-wins; duplicate IDs across files keep the first
+occurrence (uniqueness is enforced separately by
+``test_rule_id_uniqueness.py``).
+
+Reuses ``find_convention_files`` from ``test_rule_id_uniqueness.py`` so the
+discovery surface stays consistent across the two validators.
+"""
+
+from __future__ import annotations
+
+import logging
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Dict, Iterable, List, Optional, Sequence, Tuple
+
+import yaml
+
+from atdd.coach.validators.test_rule_id_uniqueness import find_convention_files
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class RuleMetadata:
+    """Normalized metadata for one declared rule.
+
+    Attributes:
+        rule_id: Stable rule identifier (e.g. ``GREEN-URN-001``).
+        convention_path: Absolute path to the ``*.convention.yaml`` that
+            declared the rule.
+        severity: Integer severity 1..5 (per SPEC-COACH-RULEID-0003) when
+            present. Some legacy rules use string severities (``"error"``);
+            those are preserved as-is.
+        description: One-line human-readable rule statement (may be empty
+            for legacy rules).
+        recipe: Optional pointer to a peer ``*.recipe.yaml`` file
+            (without the ``.recipe.yaml`` suffix).
+        introduced_in: Optional toolkit version that first published this rule.
+    """
+
+    rule_id: str
+    convention_path: Path
+    severity: object = None
+    description: str = ""
+    recipe: Optional[str] = None
+    introduced_in: Optional[str] = None
+
+
+def _build_metadata(rule_id: str, raw: Dict, path: Path) -> RuleMetadata:
+    return RuleMetadata(
+        rule_id=rule_id,
+        convention_path=path,
+        severity=raw.get("severity"),
+        description=str(raw.get("description") or ""),
+        recipe=raw.get("recipe") if isinstance(raw.get("recipe"), str) else None,
+        introduced_in=(
+            raw.get("introduced_in")
+            if isinstance(raw.get("introduced_in"), str)
+            else None
+        ),
+    )
+
+
+def _walk_node(
+    node, file_path: Path
+) -> Iterable[Tuple[str, Dict]]:
+    """Recursively yield ``(rule_id, raw_dict)`` for every shape A or B rule."""
+    if isinstance(node, dict):
+        for key, value in node.items():
+            if key == "rules":
+                yield from _extract_from_rules_block(value, file_path)
+            else:
+                yield from _walk_node(value, file_path)
+    elif isinstance(node, list):
+        for item in node:
+            yield from _walk_node(item, file_path)
+
+
+def _extract_from_rules_block(
+    value, file_path: Path
+) -> Iterable[Tuple[str, Dict]]:
+    """Yield ``(rule_id, raw_dict)`` from a single ``rules:`` block.
+
+    Shape A: list of dicts each carrying ``id:``.
+    Shape B: dict keyed by rule id (each value is a dict).
+    Shape C: dict whose value dicts have no ``id:`` field — skipped with a
+    debug log so the operator can see why entries were dropped.
+    """
+    if isinstance(value, list):
+        # Shape A. Each item is a dict with an `id` field. Recurse into the
+        # value bodies too, so nested `rules:` (e.g. green.convention.yaml's
+        # urn_naming.rules) still contribute when this list is itself nested.
+        for item in value:
+            if isinstance(item, dict):
+                rid = item.get("id")
+                if isinstance(rid, str) and rid:
+                    yield (rid, item)
+                yield from _walk_node(item, file_path)
+        return
+
+    if isinstance(value, dict):
+        skipped = 0
+        for key, body in value.items():
+            if not isinstance(body, dict):
+                continue
+            rid = body.get("id")
+            if isinstance(rid, str) and rid:
+                # Shape B: keyed mapping, body has explicit `id:`.
+                yield (rid, body)
+            elif isinstance(key, str) and _looks_like_rule_id(key):
+                # Shape B variant: key IS the rule_id, body has no `id:` field.
+                yield (key, body)
+            else:
+                # Shape C: semantic-keyed mapping with no `id:` — skip.
+                skipped += 1
+            # Recurse so deeply-nested rules: blocks still contribute.
+            yield from _walk_node(body, file_path)
+        if skipped:
+            _LOGGER.debug(
+                "rule_id_registry: skipped %d entries in %s (shape C — no id field)",
+                skipped, file_path,
+            )
+
+
+def _looks_like_rule_id(s: str) -> bool:
+    """Cheap heuristic: rule IDs are uppercase + hyphenated, never snake_case."""
+    return bool(s) and s == s.upper() and "-" in s and "_" not in s
+
+
+def _load_yaml(path: Path):
+    try:
+        with open(path) as fh:
+            return yaml.safe_load(fh)
+    except (OSError, yaml.YAMLError) as exc:
+        _LOGGER.debug("rule_id_registry: skipping malformed YAML %s: %s", path, exc)
+        return None
+
+
+def build_registry(roots: Optional[Sequence[Path]] = None) -> Dict[str, RuleMetadata]:
+    """Build the ``{rule_id: RuleMetadata}`` index.
+
+    Args:
+        roots: Optional explicit search roots. When omitted, uses the
+            toolkit-wide convention discovery from
+            ``test_rule_id_uniqueness.find_convention_files``.
+
+    Returns:
+        Dict keyed by rule_id. First occurrence wins on cross-file
+        duplication (uniqueness is enforced by
+        ``test_rule_id_uniqueness.py``).
+    """
+    if roots is None:
+        files: List[Path] = find_convention_files()
+    else:
+        files = []
+        for root in roots:
+            if not Path(root).is_dir():
+                continue
+            files.extend(sorted(Path(root).rglob("*.convention.yaml")))
+
+    registry: Dict[str, RuleMetadata] = {}
+    for path in files:
+        if "__pycache__" in path.parts:
+            continue
+        data = _load_yaml(path)
+        if data is None:
+            continue
+        for rule_id, raw in _walk_node(data, path):
+            if rule_id in registry:
+                # Cross-file duplicate. Keep the first occurrence; uniqueness
+                # is enforced separately, so we don't need to choose here.
+                continue
+            registry[rule_id] = _build_metadata(rule_id, raw, path)
+    return registry
+
+
+__all__ = ["RuleMetadata", "build_registry"]

--- a/src/atdd/coach/utils/rule_id_registry.py
+++ b/src/atdd/coach/utils/rule_id_registry.py
@@ -34,7 +34,7 @@ import yaml
 from atdd.coach.validators.test_rule_id_uniqueness import find_convention_files
 
 
-_LOGGER = logging.getLogger(__name__)
+_logger = logging.getLogger(__name__)
 
 
 @dataclass(frozen=True)
@@ -133,7 +133,7 @@ def _extract_from_rules_block(
             # Recurse so deeply-nested rules: blocks still contribute.
             yield from _walk_node(body, file_path)
         if skipped:
-            _LOGGER.debug(
+            _logger.debug(
                 "rule_id_registry: skipped %d entries in %s (shape C — no id field)",
                 skipped, file_path,
             )
@@ -149,7 +149,7 @@ def _load_yaml(path: Path):
         with open(path) as fh:
             return yaml.safe_load(fh)
     except (OSError, yaml.YAMLError) as exc:
-        _LOGGER.debug("rule_id_registry: skipping malformed YAML %s: %s", path, exc)
+        _logger.debug("rule_id_registry: skipping malformed YAML %s: %s", path, exc)
         return None
 
 

--- a/src/atdd/coach/utils/tests/test_rule_id_registry.py
+++ b/src/atdd/coach/utils/tests/test_rule_id_registry.py
@@ -1,0 +1,194 @@
+"""
+Unit tests for atdd.coach.utils.rule_id_registry.
+
+URN: urn:atdd:test:coach:utils:rule_id_registry
+Issue: #387 — Registry Coherence Validator (Phase 1: walker)
+
+The walker reads every ``src/atdd/**/conventions/*.yaml`` and normalizes the
+three observed ``rules:`` shapes into a single ``{rule_id: RuleMetadata}``
+index. Shape A and B contribute IDs; shape C is skipped.
+"""
+
+from __future__ import annotations
+
+from pathlib import Path
+
+import pytest
+
+from atdd.coach.utils.rule_id_registry import (
+    RuleMetadata,
+    build_registry,
+)
+
+
+pytestmark = [pytest.mark.coach, pytest.mark.platform]
+
+
+# ---------------------------------------------------------------------------
+# Fixtures
+# ---------------------------------------------------------------------------
+def _write(p: Path, body: str) -> Path:
+    p.parent.mkdir(parents=True, exist_ok=True)
+    p.write_text(body)
+    return p
+
+
+@pytest.fixture
+def shape_a_file(tmp_path: Path) -> Path:
+    """Shape A: rules: is a list of objects, each with `id:`."""
+    return _write(
+        tmp_path / "conventions" / "alpha.convention.yaml",
+        """
+schema_version: "1.0.0"
+convention_id: "test.alpha"
+rules:
+  - id: "GREEN-URN-001"
+    severity: 3
+    description: "Component URN required as first non-empty line"
+    introduced_in: "1.50.0"
+  - id: "GREEN-URN-002"
+    severity: 3
+    description: "URN must include layer segment"
+    recipe: "adapter"
+""",
+    )
+
+
+@pytest.fixture
+def shape_b_file(tmp_path: Path) -> Path:
+    """Shape B: rules: is a mapping keyed by rule id."""
+    return _write(
+        tmp_path / "conventions" / "bravo.convention.yaml",
+        """
+schema_version: "1.0.0"
+convention_id: "test.bravo"
+rules:
+  COACH-SILENT-SWALLOW-001:
+    id: COACH-SILENT-SWALLOW-001
+    severity: 4
+    description: "Exception handlers must log, re-raise, or otherwise observably react"
+""",
+    )
+
+
+@pytest.fixture
+def shape_c_file(tmp_path: Path) -> Path:
+    """Shape C: rules: is a semantic-keyed mapping with no `id:` field."""
+    return _write(
+        tmp_path / "conventions" / "charlie.convention.yaml",
+        """
+schema_version: "1.0.0"
+convention_id: "test.charlie"
+rules:
+  worktree_per_issue:
+    rule: "Every orchestrated issue MUST have its own git worktree"
+    anti_pattern: "Reusing a worktree across issues"
+    enforced_by: "atdd orchestrate"
+""",
+    )
+
+
+@pytest.fixture
+def mixed_shape_file(tmp_path: Path) -> Path:
+    """Shape A nested under a key, plus shape B at top-level."""
+    return _write(
+        tmp_path / "conventions" / "delta.convention.yaml",
+        """
+schema_version: "1.0.0"
+convention_id: "test.delta"
+rules:
+  COACH-MIX-001:
+    id: COACH-MIX-001
+    severity: 2
+    description: "top-level shape-B rule"
+green_phase:
+  urn_naming:
+    rules:
+      - id: "GREEN-MIX-002"
+        severity: 3
+        description: "nested shape-A rule"
+""",
+    )
+
+
+@pytest.fixture
+def malformed_yaml_file(tmp_path: Path) -> Path:
+    return _write(
+        tmp_path / "conventions" / "broken.convention.yaml",
+        "rules:\n  - id: BAD\n    description: missing severity\n  -- not yaml --\n",
+    )
+
+
+# ---------------------------------------------------------------------------
+# Tests
+# ---------------------------------------------------------------------------
+class TestBuildRegistryShapes:
+    def test_shape_a_contributes_rules(self, shape_a_file: Path):
+        registry = build_registry(roots=[shape_a_file.parent.parent])
+        assert "GREEN-URN-001" in registry
+        assert "GREEN-URN-002" in registry
+
+    def test_shape_a_metadata_normalized(self, shape_a_file: Path):
+        registry = build_registry(roots=[shape_a_file.parent.parent])
+        meta = registry["GREEN-URN-001"]
+        assert isinstance(meta, RuleMetadata)
+        assert meta.severity == 3
+        assert "URN required" in meta.description
+        assert meta.introduced_in == "1.50.0"
+        assert meta.convention_path == shape_a_file
+        # recipe is optional
+        assert registry["GREEN-URN-002"].recipe == "adapter"
+
+    def test_shape_b_contributes_rules(self, shape_b_file: Path):
+        registry = build_registry(roots=[shape_b_file.parent.parent])
+        assert "COACH-SILENT-SWALLOW-001" in registry
+        assert registry["COACH-SILENT-SWALLOW-001"].severity == 4
+
+    def test_shape_c_skipped(self, shape_c_file: Path):
+        """Shape C entries have no id: field; they must NOT pollute the registry."""
+        registry = build_registry(roots=[shape_c_file.parent.parent])
+        assert "worktree_per_issue" not in registry
+        assert registry == {}
+
+    def test_mixed_shape_file_merges_both(self, mixed_shape_file: Path):
+        registry = build_registry(roots=[mixed_shape_file.parent.parent])
+        assert "COACH-MIX-001" in registry
+        assert "GREEN-MIX-002" in registry
+
+
+class TestBuildRegistryWalking:
+    def test_walks_every_convention_file(
+        self, shape_a_file: Path, shape_b_file: Path, shape_c_file: Path,
+    ):
+        # All three live under the same conventions/ dir.
+        root = shape_a_file.parent.parent
+        registry = build_registry(roots=[root])
+        # Shape A + B contribute, shape C is skipped.
+        assert {"GREEN-URN-001", "GREEN-URN-002", "COACH-SILENT-SWALLOW-001"} <= set(registry)
+
+    def test_malformed_yaml_does_not_raise(self, malformed_yaml_file: Path):
+        # Malformed YAML must not blow up the walker — surface the path
+        # silently (or via debug log), but return a usable registry.
+        registry = build_registry(roots=[malformed_yaml_file.parent.parent])
+        assert isinstance(registry, dict)
+
+
+class TestBuildRegistryAgainstRealConventions:
+    """Smoke test against the real toolkit conventions tree."""
+
+    def test_real_registry_contains_known_ids(self):
+        registry = build_registry()
+        # Substrate IDs declared in the toolkit:
+        # green.convention.yaml ships GREEN-URN-001 (shape A)
+        # logging.convention.yaml ships COACH-SILENT-SWALLOW-001 (shape A)
+        assert "COACH-SILENT-SWALLOW-001" in registry, (
+            f"expected COACH-SILENT-SWALLOW-001 in real registry; "
+            f"got {sorted(registry)[:5]}..."
+        )
+
+    def test_real_registry_metadata_is_RuleMetadata(self):
+        registry = build_registry()
+        for rule_id, meta in registry.items():
+            assert isinstance(meta, RuleMetadata), f"{rule_id}: {type(meta)}"
+            assert isinstance(meta.convention_path, Path)
+            assert isinstance(meta.description, str)

--- a/src/atdd/coach/validators/rule_id_emission_extractor.py
+++ b/src/atdd/coach/validators/rule_id_emission_extractor.py
@@ -39,10 +39,14 @@ live outside the scan roots.
 
 from __future__ import annotations
 
+import logging
 import re
 from dataclasses import dataclass
 from pathlib import Path
 from typing import Iterable, List, Sequence
+
+
+_logger = logging.getLogger(__name__)
 
 
 # ---------------------------------------------------------------------------
@@ -80,7 +84,11 @@ def extract_emissions(file_path: Path) -> Iterable[Emission]:
     """
     try:
         text = file_path.read_text(encoding="utf-8")
-    except (OSError, UnicodeDecodeError):
+    except (OSError, UnicodeDecodeError) as exc:
+        _logger.debug(
+            "rule_id_emission_extractor: skipping unreadable file %s: %s",
+            file_path, exc,
+        )
         return
 
     seen = set()

--- a/src/atdd/coach/validators/rule_id_emission_extractor.py
+++ b/src/atdd/coach/validators/rule_id_emission_extractor.py
@@ -88,6 +88,7 @@ def extract_emissions(file_path: Path) -> Iterable[Emission]:
         _logger.debug(
             "rule_id_emission_extractor: skipping unreadable file %s: %s",
             file_path, exc,
+            extra={"file": str(file_path), "error_type": type(exc).__name__},
         )
         return
 

--- a/src/atdd/coach/validators/rule_id_emission_extractor.py
+++ b/src/atdd/coach/validators/rule_id_emission_extractor.py
@@ -1,0 +1,144 @@
+# URN: component:govern-lifecycle:enforcement-substrate:rule_id_emission_extractor:backend:domain
+# Runtime: python
+# Purpose: Regex-based extractor for rule_id emissions in production validator + command source.
+
+"""
+Rule-ID emission extractor (issue #387, Phase 2 helper).
+
+Extracts every string-literal rule_id emission from a Python source file.
+Decision #1 of #387 fixes the regex set:
+
+  Pattern A — `Violation\\(\\s*rule_id\\s*=\\s*"([A-Z][A-Z0-9-]+)"`
+              Matches the canonical Violation kwarg form.
+  Pattern B — `\\b([A-Z][A-Z0-9_]*)\\s*=\\s*"([A-Z][A-Z0-9]*(?:-[A-Z0-9]+){2,})"`
+              Matches any all-caps identifier whose VALUE conforms to the
+              rule_id grammar — catches RULE_ID_*, *_RULE_ID, AND constants
+              whose names contain neither (RULE_EMPTY_RENDER, XSS_RULE_ID,
+              RULE_DYNAMIC_TRAIN_ID, RULE_ALLOWLIST_MIGRATION, …).
+  Pattern C — catch-all for any keyword-arg string literal that names
+              ``rule_id`` (e.g. babysit's auto-approval builder); ensures a
+              non-Violation constructor still gets scanned.
+
+Out-of-scope for v1: dynamic emissions where the rhs is an attribute or
+subscript reference instead of a string literal (e.g. babysit's per-rule
+loop where the rhs reads from a parsed config). These require AST +
+control-flow analysis and are deferred until v1's false-negative rate
+forces the upgrade.
+
+Decision #4 / Success Criteria file exclusions are applied in
+``iter_scan_files``:
+  - ``**/tests/**`` (pytest unit tests of validators)
+  - ``**/fixtures/**`` (placeholder rule_ids in fixtures)
+
+Validator files named ``test_*.py`` are NOT excluded — those files ARE the
+production validators (pytest discovers validators by filename). The scan
+roots (``validators/``, ``coach/commands/``) bound the surface; stragglers
+like ``coder/baselines/test_ratchet.py`` are out of reach because they
+live outside the scan roots.
+"""
+
+from __future__ import annotations
+
+import re
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Iterable, List, Sequence
+
+
+# ---------------------------------------------------------------------------
+# Decision #1 regex set
+# ---------------------------------------------------------------------------
+_PATTERN_A = re.compile(
+    r"""Violation\(\s*rule_id\s*=\s*"([A-Z][A-Z0-9\-]+)" """.strip()
+)
+_PATTERN_B = re.compile(
+    r"""\b([A-Z][A-Z0-9_]*)\s*=\s*"([A-Z][A-Z0-9]*(?:-[A-Z0-9]+){2,})" """.strip()
+)
+_PATTERN_C = re.compile(
+    r"""\brule_id\s*=\s*"([A-Z][A-Z0-9\-]+)" """.strip()
+)
+
+EMISSION_PATTERNS = (_PATTERN_A, _PATTERN_B, _PATTERN_C)
+
+
+@dataclass(frozen=True)
+class Emission:
+    """One occurrence of a rule_id string literal in production source."""
+
+    rule_id: str
+    file_path: Path
+    line: int
+
+
+def extract_emissions(file_path: Path) -> Iterable[Emission]:
+    """Yield every rule_id emission from *file_path*.
+
+    Multiple patterns may match the same line; emissions are deduplicated by
+    ``(file_path, line, rule_id)`` so a single ``RULE_ID = "<grammar>"``
+    constant (matched by both pattern B and pattern C in slightly different
+    scopes) only surfaces once.
+    """
+    try:
+        text = file_path.read_text(encoding="utf-8")
+    except (OSError, UnicodeDecodeError):
+        return
+
+    seen = set()
+    for lineno, line in enumerate(text.splitlines(), start=1):
+        for match in _PATTERN_A.finditer(line):
+            rid = match.group(1)
+            key = (str(file_path), lineno, rid)
+            if key not in seen:
+                seen.add(key)
+                yield Emission(rule_id=rid, file_path=file_path, line=lineno)
+        for match in _PATTERN_B.finditer(line):
+            # group(1) = identifier, group(2) = rule_id grammar value
+            rid = match.group(2)
+            key = (str(file_path), lineno, rid)
+            if key not in seen:
+                seen.add(key)
+                yield Emission(rule_id=rid, file_path=file_path, line=lineno)
+        for match in _PATTERN_C.finditer(line):
+            rid = match.group(1)
+            key = (str(file_path), lineno, rid)
+            if key not in seen:
+                seen.add(key)
+                yield Emission(rule_id=rid, file_path=file_path, line=lineno)
+
+
+def iter_scan_files(roots: Sequence[Path]) -> Iterable[Path]:
+    """Yield every production *.py file under *roots*.
+
+    Excluded:
+      - any file inside a ``tests/`` or ``fixtures/`` directory (anywhere)
+      - ``__init__.py`` (no rule_id emissions are declared in package init)
+      - ``__pycache__/`` artifacts
+
+    Validator files named ``test_*.py`` directly under a scan root ARE
+    included — those files are the production validators.
+    """
+    seen: set = set()
+    for root in roots:
+        if not Path(root).is_dir():
+            continue
+        for path in Path(root).rglob("*.py"):
+            parts = set(path.parts)
+            if "tests" in parts or "fixtures" in parts:
+                continue
+            if "__pycache__" in parts:
+                continue
+            if path.name == "__init__.py":
+                continue
+            key = str(path.resolve())
+            if key in seen:
+                continue
+            seen.add(key)
+            yield path
+
+
+__all__ = [
+    "EMISSION_PATTERNS",
+    "Emission",
+    "extract_emissions",
+    "iter_scan_files",
+]

--- a/src/atdd/coach/validators/test_rule_id_registry_coherence.py
+++ b/src/atdd/coach/validators/test_rule_id_registry_coherence.py
@@ -9,20 +9,25 @@ Walks production validator + command source, extracts rule_id emissions via
 three regex patterns (Decision #1), and cross-checks against the registry
 built from every ``*.convention.yaml``. Drift surfaces as plain text:
 
-  - permissive mode (default)     → WARN, exit 0
-  - ``--strict-coherence``        → ERROR, exit 1
+  - permissive mode (default)     -> WARN, exit 0
+  - ``ATDD_STRICT_COHERENCE=1``   -> ERROR, exit 1
+
+The CLI flag ``--strict-coherence`` on ``atdd validate coach`` sets the
+env var.
 
 Decision #2: this validator does NOT emit ``Violation`` records for drift —
 drift is a config issue, not a runtime violation. Plain text avoids
 meta-recursion (a validator emitting violations about emissions).
+
+Helper unit tests live under ``validators/tests/`` so their fixture string
+literals are not picked up by this validator's own scan.
 """
 
 from __future__ import annotations
 
 import os
-import re
 from pathlib import Path
-from typing import Iterable, List, Tuple
+from typing import List, Tuple
 
 import pytest
 
@@ -30,8 +35,6 @@ import atdd
 from atdd.coach.utils.repo import find_repo_root
 from atdd.coach.utils.rule_id_registry import build_registry
 from atdd.coach.validators.rule_id_emission_extractor import (
-    EMISSION_PATTERNS,
-    Emission,
     extract_emissions,
     iter_scan_files,
 )
@@ -45,9 +48,9 @@ pytestmark = [pytest.mark.coach, pytest.mark.platform]
 # ---------------------------------------------------------------------------
 ATDD_PKG_DIR = Path(atdd.__file__).resolve().parent
 
-# Production scan surface (Decision #4):
+# Production scan surface:
 #   validators/**/*.py + coach/commands/**/*.py
-#   exclude:  **/tests/**, **/fixtures/**, test_*.py
+#   exclude:  **/tests/**, **/fixtures/**
 _SCAN_ROOTS_REL = (
     ("coach", "validators"),
     ("coder", "validators"),
@@ -61,155 +64,19 @@ def _scan_roots() -> List[Path]:
     """Return absolute paths to every production scan root (toolkit + repo checkout)."""
     repo_src = find_repo_root() / "src" / "atdd"
     out: List[Path] = []
+    seen = set()
     for parts in _SCAN_ROOTS_REL:
         for base in (ATDD_PKG_DIR, repo_src):
             cand = base.joinpath(*parts)
-            if cand.is_dir():
+            key = str(cand.resolve())
+            if cand.is_dir() and key not in seen:
+                seen.add(key)
                 out.append(cand)
     return out
 
 
-# ---------------------------------------------------------------------------
-# Tests
-# ---------------------------------------------------------------------------
-class TestEmissionExtractorPatterns:
-    """The three Decision #1 regex patterns extract the expected shapes."""
-
-    def test_pattern_a_matches_violation_kwarg(self, tmp_path: Path):
-        f = tmp_path / "v.py"
-        f.write_text(
-            'from x import Violation\n'
-            'v = Violation(rule_id="COACH-PRGATE-0003", severity=4, location="a:1", detail="d")\n'
-        )
-        emissions = list(extract_emissions(f))
-        ids = {e.rule_id for e in emissions}
-        assert "COACH-PRGATE-0003" in ids
-
-    def test_pattern_b_matches_constant_with_RULE_ID_prefix(self, tmp_path: Path):
-        f = tmp_path / "c.py"
-        f.write_text('RULE_ID_PRGATE_GREEN = "COACH-PRGATE-0003"\n')
-        emissions = list(extract_emissions(f))
-        assert any(e.rule_id == "COACH-PRGATE-0003" for e in emissions)
-
-    def test_pattern_b_matches_constant_without_RULE_ID_prefix(self, tmp_path: Path):
-        """Decision #1: pattern (b) matches by VALUE shape, not LHS shape.
-
-        These constants must be caught even though the LHS does not contain
-        `RULE_ID`: RULE_EMPTY_RENDER, RULE_HARNESS_ERROR, XSS_RULE_ID,
-        RULE_DYNAMIC_TRAIN_ID, RULE_ALLOWLIST_MIGRATION.
-        """
-        f = tmp_path / "c2.py"
-        f.write_text(
-            'RULE_EMPTY_RENDER = "TESTER-RENDER-001"\n'
-            'XSS_RULE_ID = "SECURITY-XSS-001"\n'
-            'RULE_DYNAMIC_TRAIN_ID = "BOUNDARIES-ROUTE-COVERAGE-003"\n'
-        )
-        ids = {e.rule_id for e in extract_emissions(f)}
-        assert "TESTER-RENDER-001" in ids
-        assert "SECURITY-XSS-001" in ids
-        assert "BOUNDARIES-ROUTE-COVERAGE-003" in ids
-
-    def test_pattern_c_matches_keyword_arg_in_other_constructors(self, tmp_path: Path):
-        """Catch-all `\\brule_id\\s*=\\s*"..."` (e.g. in non-Violation constructors)."""
-        f = tmp_path / "c3.py"
-        f.write_text(
-            'event = build(rule_id="COACH-BABYSIT-001", reason="x")\n'
-        )
-        ids = {e.rule_id for e in extract_emissions(f)}
-        assert "COACH-BABYSIT-001" in ids
-
-    def test_dynamic_emissions_out_of_scope(self, tmp_path: Path):
-        """Decision #1: ``rule_id=rule["id"]`` and ``rule_id=p.rule_id`` are
-        deferred — they require AST + control-flow analysis."""
-        f = tmp_path / "dyn.py"
-        f.write_text(
-            'event = build(rule_id=rule["id"], reason="x")\n'
-            'event = build(rule_id=p.rule_id, reason="x")\n'
-        )
-        ids = {e.rule_id for e in extract_emissions(f)}
-        assert ids == set()
-
-    def test_emission_records_file_and_line(self, tmp_path: Path):
-        f = tmp_path / "loc.py"
-        f.write_text(
-            "# header\n"
-            'RULE_X = "COACH-PRGATE-0003"\n'
-        )
-        emissions = list(extract_emissions(f))
-        assert len(emissions) == 1
-        e = emissions[0]
-        assert e.file_path == f
-        assert e.line == 2
-
-
-class TestScanSurfaceExclusions:
-    """Decision #4: tests, fixtures, and test_*.py files (anywhere) are excluded."""
-
-    def test_skips_tests_directory(self, tmp_path: Path):
-        root = tmp_path / "validators"
-        (root / "tests").mkdir(parents=True)
-        (root / "tests" / "test_inner.py").write_text(
-            'RULE_X = "COACH-PRGATE-0003"\n'
-        )
-        files = list(iter_scan_files([root]))
-        assert all("tests" not in p.parts for p in files)
-
-    def test_skips_fixtures_directory(self, tmp_path: Path):
-        root = tmp_path / "validators"
-        (root / "fixtures").mkdir(parents=True)
-        (root / "fixtures" / "sample.py").write_text(
-            'RULE_X = "COACH-PRGATE-0003"\n'
-        )
-        files = list(iter_scan_files([root]))
-        assert all("fixtures" not in p.parts for p in files)
-
-    def test_skips_test_underscore_files_anywhere(self, tmp_path: Path):
-        """Pytest discovers tests by filename, not directory; exclusion mirrors that."""
-        root = tmp_path / "validators"
-        sub = root / "baselines"
-        sub.mkdir(parents=True)
-        (sub / "test_ratchet.py").write_text(
-            'RULE_X = "COACH-PRGATE-0003"\n'
-        )
-        (sub / "ratchet.py").write_text(
-            'RULE_X = "COACH-PRGATE-0003"\n'
-        )
-        files = list(iter_scan_files([root]))
-        names = {p.name for p in files}
-        assert "ratchet.py" in names
-        assert "test_ratchet.py" not in names
-
-
-class TestCoherenceValidatorAgainstCurrentMain:
-    """Run against the real toolkit source — at least one drift entry must surface."""
-
-    def test_known_unregistered_id_surfaces(self):
-        """The seeded drift example from #387: ``COACH-PRGATE-0003`` is declared
-        in test_pr_phase_alignment.py but not in any convention rules: block.
-        """
-        registry = build_registry()
-        roots = _scan_roots()
-        assert roots, "no production scan roots resolved"
-        unregistered: List[Tuple[Path, int, str]] = []
-        for f in iter_scan_files(roots):
-            for e in extract_emissions(f):
-                if e.rule_id not in registry:
-                    unregistered.append((e.file_path, e.line, e.rule_id))
-        ids = {rid for (_, _, rid) in unregistered}
-        # Confirmed seeds from the issue body:
-        assert "COACH-PRGATE-0003" in ids
-
-
-@pytest.mark.coach
-def test_rule_id_registry_coherence():
-    """Permissive-mode coherence check.
-
-    Default behavior: emit a WARN block listing every emission whose
-    rule_id is missing from the registry, then PASS (exit 0).
-
-    Strict mode is opt-in via ``ATDD_STRICT_COHERENCE=1`` (the env var the
-    CLI flag `--strict-coherence` sets) — when enabled, drift fails the test.
-    """
+def _collect_drift() -> List[Tuple[Path, int, str]]:
+    """Return ``[(file_path, line, rule_id), ...]`` for unregistered emissions."""
     registry = build_registry()
     roots = _scan_roots()
     drift: List[Tuple[Path, int, str]] = []
@@ -217,10 +84,10 @@ def test_rule_id_registry_coherence():
         for e in extract_emissions(f):
             if e.rule_id not in registry:
                 drift.append((e.file_path, e.line, e.rule_id))
+    return drift
 
-    if not drift:
-        return  # registry is fully coherent; nothing to surface
 
+def _format_drift(drift: List[Tuple[Path, int, str]]) -> str:
     repo_root = find_repo_root()
     lines = [
         f"[WARN] rule_id_registry_coherence: "
@@ -233,20 +100,55 @@ def test_rule_id_registry_coherence():
             rel = fp
         lines.append(f"  {rel}:{ln}   {rid}   not in any convention rules: block")
     lines.append("  Run with --strict-coherence to fail CI on this.")
+    return "\n".join(lines)
 
-    msg = "\n".join(lines)
 
-    strict = os.environ.get("ATDD_STRICT_COHERENCE") == "1"
-    if strict:
-        pytest.fail(msg)
+# ---------------------------------------------------------------------------
+# Tests / validator entry points
+# ---------------------------------------------------------------------------
+class TestCoherenceValidatorAgainstCurrentMain:
+    """Smoke assertions against the live toolkit source."""
+
+    def test_seeded_drift_id_surfaces(self):
+        """The seeded drift example from #387: ``COACH-PRGATE-0003`` is
+        declared in ``test_pr_phase_alignment.py`` but not in any convention
+        rules: block.
+        """
+        roots = _scan_roots()
+        assert roots, "no production scan roots resolved"
+
+        registry = build_registry()
+        unregistered = set()
+        for f in iter_scan_files(roots):
+            for e in extract_emissions(f):
+                if e.rule_id not in registry:
+                    unregistered.add(e.rule_id)
+
+        seeded = "COACH-" + "PRGATE-0003"
+        assert seeded in unregistered, (
+            f"expected seed {seeded} in drift; got {sorted(unregistered)[:10]}..."
+        )
+
+
+@pytest.mark.coach
+def test_rule_id_registry_coherence():
+    """Permissive-mode coherence check.
+
+    Default behavior: print a WARN block listing every emission whose
+    rule_id is missing from the registry, then PASS (exit 0).
+
+    Strict mode is opt-in via ``ATDD_STRICT_COHERENCE=1`` (the env var the
+    CLI flag ``--strict-coherence`` sets) — when enabled, drift fails the
+    test, which makes ``atdd validate coach`` exit 1.
+    """
+    drift = _collect_drift()
+    if not drift:
+        return  # registry is fully coherent; nothing to surface
+
+    msg = _format_drift(drift)
+
+    if os.environ.get("ATDD_STRICT_COHERENCE") == "1":
+        pytest.fail(msg.replace("[WARN]", "[ERROR]"))
     else:
         # Plain-text WARN — visible in pytest -v output but does not fail.
         print("\n" + msg)
-
-
-__all__ = [
-    "EMISSION_PATTERNS",
-    "Emission",
-    "extract_emissions",
-    "iter_scan_files",
-]

--- a/src/atdd/coach/validators/test_rule_id_registry_coherence.py
+++ b/src/atdd/coach/validators/test_rule_id_registry_coherence.py
@@ -25,6 +25,7 @@ literals are not picked up by this validator's own scan.
 
 from __future__ import annotations
 
+import logging
 import os
 from pathlib import Path
 from typing import List, Tuple
@@ -38,6 +39,9 @@ from atdd.coach.validators.rule_id_emission_extractor import (
     extract_emissions,
     iter_scan_files,
 )
+
+
+_logger = logging.getLogger(__name__)
 
 
 pytestmark = [pytest.mark.coach, pytest.mark.platform]
@@ -150,5 +154,15 @@ def test_rule_id_registry_coherence():
     if os.environ.get("ATDD_STRICT_COHERENCE") == "1":
         pytest.fail(msg.replace("[WARN]", "[ERROR]"))
     else:
-        # Plain-text WARN — visible in pytest -v output but does not fail.
-        print("\n" + msg)
+        # Permissive-mode WARN — surfaces drift without failing the gate.
+        _logger.warning(
+            "rule_id_registry_coherence drift: %d unregistered emission(s)",
+            len(drift),
+            extra={"drift_count": len(drift), "validator": "rule_id_registry_coherence"},
+        )
+        for fp, ln, rid in sorted(drift):
+            _logger.warning(
+                "rule_id_registry_coherence: %s:%d %s not in any convention rules: block",
+                fp, ln, rid,
+                extra={"file": str(fp), "line": ln, "rule_id": rid},
+            )

--- a/src/atdd/coach/validators/test_rule_id_registry_coherence.py
+++ b/src/atdd/coach/validators/test_rule_id_registry_coherence.py
@@ -1,0 +1,252 @@
+# URN: component:govern-lifecycle:enforcement-substrate:rule_id_registry_coherence:backend:domain
+# Runtime: python
+# Purpose: Surface emissions whose rule_id is not declared in any convention rules: block.
+
+"""
+Coach validator for rule-id registry coherence (issue #387).
+
+Walks production validator + command source, extracts rule_id emissions via
+three regex patterns (Decision #1), and cross-checks against the registry
+built from every ``*.convention.yaml``. Drift surfaces as plain text:
+
+  - permissive mode (default)     → WARN, exit 0
+  - ``--strict-coherence``        → ERROR, exit 1
+
+Decision #2: this validator does NOT emit ``Violation`` records for drift —
+drift is a config issue, not a runtime violation. Plain text avoids
+meta-recursion (a validator emitting violations about emissions).
+"""
+
+from __future__ import annotations
+
+import os
+import re
+from pathlib import Path
+from typing import Iterable, List, Tuple
+
+import pytest
+
+import atdd
+from atdd.coach.utils.repo import find_repo_root
+from atdd.coach.utils.rule_id_registry import build_registry
+from atdd.coach.validators.rule_id_emission_extractor import (
+    EMISSION_PATTERNS,
+    Emission,
+    extract_emissions,
+    iter_scan_files,
+)
+
+
+pytestmark = [pytest.mark.coach, pytest.mark.platform]
+
+
+# ---------------------------------------------------------------------------
+# Path constants
+# ---------------------------------------------------------------------------
+ATDD_PKG_DIR = Path(atdd.__file__).resolve().parent
+
+# Production scan surface (Decision #4):
+#   validators/**/*.py + coach/commands/**/*.py
+#   exclude:  **/tests/**, **/fixtures/**, test_*.py
+_SCAN_ROOTS_REL = (
+    ("coach", "validators"),
+    ("coder", "validators"),
+    ("tester", "validators"),
+    ("planner", "validators"),
+    ("coach", "commands"),
+)
+
+
+def _scan_roots() -> List[Path]:
+    """Return absolute paths to every production scan root (toolkit + repo checkout)."""
+    repo_src = find_repo_root() / "src" / "atdd"
+    out: List[Path] = []
+    for parts in _SCAN_ROOTS_REL:
+        for base in (ATDD_PKG_DIR, repo_src):
+            cand = base.joinpath(*parts)
+            if cand.is_dir():
+                out.append(cand)
+    return out
+
+
+# ---------------------------------------------------------------------------
+# Tests
+# ---------------------------------------------------------------------------
+class TestEmissionExtractorPatterns:
+    """The three Decision #1 regex patterns extract the expected shapes."""
+
+    def test_pattern_a_matches_violation_kwarg(self, tmp_path: Path):
+        f = tmp_path / "v.py"
+        f.write_text(
+            'from x import Violation\n'
+            'v = Violation(rule_id="COACH-PRGATE-0003", severity=4, location="a:1", detail="d")\n'
+        )
+        emissions = list(extract_emissions(f))
+        ids = {e.rule_id for e in emissions}
+        assert "COACH-PRGATE-0003" in ids
+
+    def test_pattern_b_matches_constant_with_RULE_ID_prefix(self, tmp_path: Path):
+        f = tmp_path / "c.py"
+        f.write_text('RULE_ID_PRGATE_GREEN = "COACH-PRGATE-0003"\n')
+        emissions = list(extract_emissions(f))
+        assert any(e.rule_id == "COACH-PRGATE-0003" for e in emissions)
+
+    def test_pattern_b_matches_constant_without_RULE_ID_prefix(self, tmp_path: Path):
+        """Decision #1: pattern (b) matches by VALUE shape, not LHS shape.
+
+        These constants must be caught even though the LHS does not contain
+        `RULE_ID`: RULE_EMPTY_RENDER, RULE_HARNESS_ERROR, XSS_RULE_ID,
+        RULE_DYNAMIC_TRAIN_ID, RULE_ALLOWLIST_MIGRATION.
+        """
+        f = tmp_path / "c2.py"
+        f.write_text(
+            'RULE_EMPTY_RENDER = "TESTER-RENDER-001"\n'
+            'XSS_RULE_ID = "SECURITY-XSS-001"\n'
+            'RULE_DYNAMIC_TRAIN_ID = "BOUNDARIES-ROUTE-COVERAGE-003"\n'
+        )
+        ids = {e.rule_id for e in extract_emissions(f)}
+        assert "TESTER-RENDER-001" in ids
+        assert "SECURITY-XSS-001" in ids
+        assert "BOUNDARIES-ROUTE-COVERAGE-003" in ids
+
+    def test_pattern_c_matches_keyword_arg_in_other_constructors(self, tmp_path: Path):
+        """Catch-all `\\brule_id\\s*=\\s*"..."` (e.g. in non-Violation constructors)."""
+        f = tmp_path / "c3.py"
+        f.write_text(
+            'event = build(rule_id="COACH-BABYSIT-001", reason="x")\n'
+        )
+        ids = {e.rule_id for e in extract_emissions(f)}
+        assert "COACH-BABYSIT-001" in ids
+
+    def test_dynamic_emissions_out_of_scope(self, tmp_path: Path):
+        """Decision #1: ``rule_id=rule["id"]`` and ``rule_id=p.rule_id`` are
+        deferred — they require AST + control-flow analysis."""
+        f = tmp_path / "dyn.py"
+        f.write_text(
+            'event = build(rule_id=rule["id"], reason="x")\n'
+            'event = build(rule_id=p.rule_id, reason="x")\n'
+        )
+        ids = {e.rule_id for e in extract_emissions(f)}
+        assert ids == set()
+
+    def test_emission_records_file_and_line(self, tmp_path: Path):
+        f = tmp_path / "loc.py"
+        f.write_text(
+            "# header\n"
+            'RULE_X = "COACH-PRGATE-0003"\n'
+        )
+        emissions = list(extract_emissions(f))
+        assert len(emissions) == 1
+        e = emissions[0]
+        assert e.file_path == f
+        assert e.line == 2
+
+
+class TestScanSurfaceExclusions:
+    """Decision #4: tests, fixtures, and test_*.py files (anywhere) are excluded."""
+
+    def test_skips_tests_directory(self, tmp_path: Path):
+        root = tmp_path / "validators"
+        (root / "tests").mkdir(parents=True)
+        (root / "tests" / "test_inner.py").write_text(
+            'RULE_X = "COACH-PRGATE-0003"\n'
+        )
+        files = list(iter_scan_files([root]))
+        assert all("tests" not in p.parts for p in files)
+
+    def test_skips_fixtures_directory(self, tmp_path: Path):
+        root = tmp_path / "validators"
+        (root / "fixtures").mkdir(parents=True)
+        (root / "fixtures" / "sample.py").write_text(
+            'RULE_X = "COACH-PRGATE-0003"\n'
+        )
+        files = list(iter_scan_files([root]))
+        assert all("fixtures" not in p.parts for p in files)
+
+    def test_skips_test_underscore_files_anywhere(self, tmp_path: Path):
+        """Pytest discovers tests by filename, not directory; exclusion mirrors that."""
+        root = tmp_path / "validators"
+        sub = root / "baselines"
+        sub.mkdir(parents=True)
+        (sub / "test_ratchet.py").write_text(
+            'RULE_X = "COACH-PRGATE-0003"\n'
+        )
+        (sub / "ratchet.py").write_text(
+            'RULE_X = "COACH-PRGATE-0003"\n'
+        )
+        files = list(iter_scan_files([root]))
+        names = {p.name for p in files}
+        assert "ratchet.py" in names
+        assert "test_ratchet.py" not in names
+
+
+class TestCoherenceValidatorAgainstCurrentMain:
+    """Run against the real toolkit source — at least one drift entry must surface."""
+
+    def test_known_unregistered_id_surfaces(self):
+        """The seeded drift example from #387: ``COACH-PRGATE-0003`` is declared
+        in test_pr_phase_alignment.py but not in any convention rules: block.
+        """
+        registry = build_registry()
+        roots = _scan_roots()
+        assert roots, "no production scan roots resolved"
+        unregistered: List[Tuple[Path, int, str]] = []
+        for f in iter_scan_files(roots):
+            for e in extract_emissions(f):
+                if e.rule_id not in registry:
+                    unregistered.append((e.file_path, e.line, e.rule_id))
+        ids = {rid for (_, _, rid) in unregistered}
+        # Confirmed seeds from the issue body:
+        assert "COACH-PRGATE-0003" in ids
+
+
+@pytest.mark.coach
+def test_rule_id_registry_coherence():
+    """Permissive-mode coherence check.
+
+    Default behavior: emit a WARN block listing every emission whose
+    rule_id is missing from the registry, then PASS (exit 0).
+
+    Strict mode is opt-in via ``ATDD_STRICT_COHERENCE=1`` (the env var the
+    CLI flag `--strict-coherence` sets) — when enabled, drift fails the test.
+    """
+    registry = build_registry()
+    roots = _scan_roots()
+    drift: List[Tuple[Path, int, str]] = []
+    for f in iter_scan_files(roots):
+        for e in extract_emissions(f):
+            if e.rule_id not in registry:
+                drift.append((e.file_path, e.line, e.rule_id))
+
+    if not drift:
+        return  # registry is fully coherent; nothing to surface
+
+    repo_root = find_repo_root()
+    lines = [
+        f"[WARN] rule_id_registry_coherence: "
+        f"{len(drift)} emission(s) reference unregistered rule_id(s):"
+    ]
+    for fp, ln, rid in sorted(drift):
+        try:
+            rel = fp.resolve().relative_to(repo_root.resolve())
+        except ValueError:
+            rel = fp
+        lines.append(f"  {rel}:{ln}   {rid}   not in any convention rules: block")
+    lines.append("  Run with --strict-coherence to fail CI on this.")
+
+    msg = "\n".join(lines)
+
+    strict = os.environ.get("ATDD_STRICT_COHERENCE") == "1"
+    if strict:
+        pytest.fail(msg)
+    else:
+        # Plain-text WARN — visible in pytest -v output but does not fail.
+        print("\n" + msg)
+
+
+__all__ = [
+    "EMISSION_PATTERNS",
+    "Emission",
+    "extract_emissions",
+    "iter_scan_files",
+]

--- a/src/atdd/coach/validators/tests/test_rule_id_emission_extractor.py
+++ b/src/atdd/coach/validators/tests/test_rule_id_emission_extractor.py
@@ -1,0 +1,131 @@
+"""
+Unit tests for atdd.coach.validators.rule_id_emission_extractor.
+
+URN: urn:atdd:test:coach:validators:rule_id_emission_extractor
+Issue: #387 — Registry Coherence Validator (Phase 2 helper)
+
+Lives under validators/tests/ so the rule_id literals embedded in test
+fixtures don't get scanned by the coherence validator itself.
+"""
+
+from __future__ import annotations
+
+from pathlib import Path
+
+import pytest
+
+from atdd.coach.validators.rule_id_emission_extractor import (
+    Emission,
+    extract_emissions,
+    iter_scan_files,
+)
+
+
+pytestmark = [pytest.mark.coach, pytest.mark.platform]
+
+
+class TestEmissionExtractorPatterns:
+    """The three Decision #1 regex patterns extract the expected shapes."""
+
+    def test_pattern_a_matches_violation_kwarg(self, tmp_path: Path):
+        f = tmp_path / "v.py"
+        f.write_text(
+            'from x import Violation\n'
+            'v = Violation(rule_id="COACH-PRGATE-0003", severity=4, location="a:1", detail="d")\n'
+        )
+        emissions = list(extract_emissions(f))
+        ids = {e.rule_id for e in emissions}
+        assert "COACH-PRGATE-0003" in ids
+
+    def test_pattern_b_matches_constant_with_rule_id_prefix(self, tmp_path: Path):
+        f = tmp_path / "c.py"
+        f.write_text('RULE_ID_PRGATE_GREEN = "COACH-PRGATE-0003"\n')
+        emissions = list(extract_emissions(f))
+        assert any(e.rule_id == "COACH-PRGATE-0003" for e in emissions)
+
+    def test_pattern_b_matches_constant_without_rule_id_prefix(self, tmp_path: Path):
+        """Decision #1: pattern (b) matches by VALUE shape, not LHS shape.
+
+        These constants must be caught even though the LHS does not contain
+        `RULE_ID`: RULE_EMPTY_RENDER, RULE_HARNESS_ERROR, XSS_RULE_ID,
+        RULE_DYNAMIC_TRAIN_ID, RULE_ALLOWLIST_MIGRATION.
+        """
+        f = tmp_path / "c2.py"
+        f.write_text(
+            'RULE_EMPTY_RENDER = "TESTER-RENDER-001"\n'
+            'XSS_RULE_ID = "SECURITY-XSS-001"\n'
+            'RULE_DYNAMIC_TRAIN_ID = "BOUNDARIES-ROUTE-COVERAGE-003"\n'
+        )
+        ids = {e.rule_id for e in extract_emissions(f)}
+        assert "TESTER-RENDER-001" in ids
+        assert "SECURITY-XSS-001" in ids
+        assert "BOUNDARIES-ROUTE-COVERAGE-003" in ids
+
+    def test_pattern_c_matches_keyword_arg_in_other_constructors(self, tmp_path: Path):
+        """Catch-all `\\brule_id\\s*=\\s*"..."` (e.g. in non-Violation constructors)."""
+        f = tmp_path / "c3.py"
+        f.write_text(
+            'event = build(rule_id="COACH-BABYSIT-001", reason="x")\n'
+        )
+        ids = {e.rule_id for e in extract_emissions(f)}
+        assert "COACH-BABYSIT-001" in ids
+
+    def test_dynamic_emissions_out_of_scope(self, tmp_path: Path):
+        """Decision #1: ``rule_id=rule["id"]`` and ``rule_id=p.rule_id`` are
+        deferred — they require AST + control-flow analysis."""
+        f = tmp_path / "dyn.py"
+        f.write_text(
+            'event = build(rule_id=rule["id"], reason="x")\n'
+            'event = build(rule_id=p.rule_id, reason="x")\n'
+        )
+        ids = {e.rule_id for e in extract_emissions(f)}
+        assert ids == set()
+
+    def test_emission_records_file_and_line(self, tmp_path: Path):
+        f = tmp_path / "loc.py"
+        f.write_text(
+            "# header\n"
+            'RULE_X = "COACH-PRGATE-0003"\n'
+        )
+        emissions = list(extract_emissions(f))
+        assert len(emissions) == 1
+        e = emissions[0]
+        assert isinstance(e, Emission)
+        assert e.file_path == f
+        assert e.line == 2
+
+
+class TestScanSurfaceExclusions:
+    """Success Criteria: scan ``validators/**/*.py`` + ``coach/commands/**/*.py``;
+    exclude ``**/tests/**`` and ``**/fixtures/**``."""
+
+    def test_skips_tests_directory(self, tmp_path: Path):
+        root = tmp_path / "validators"
+        (root / "tests").mkdir(parents=True)
+        (root / "tests" / "test_inner.py").write_text(
+            'RULE_X = "COACH-PRGATE-0003"\n'
+        )
+        files = list(iter_scan_files([root]))
+        assert all("tests" not in p.parts for p in files)
+
+    def test_skips_fixtures_directory(self, tmp_path: Path):
+        root = tmp_path / "validators"
+        (root / "fixtures").mkdir(parents=True)
+        (root / "fixtures" / "sample.py").write_text(
+            'RULE_X = "COACH-PRGATE-0003"\n'
+        )
+        files = list(iter_scan_files([root]))
+        assert all("fixtures" not in p.parts for p in files)
+
+    def test_includes_validator_files_directly_under_root(self, tmp_path: Path):
+        """Validator files are named test_*.py and ARE the production code."""
+        root = tmp_path / "validators"
+        root.mkdir(parents=True)
+        (root / "test_pr_phase_alignment.py").write_text(
+            'RULE_X = "COACH-PRGATE-0003"\n'
+        )
+        (root / "_violation.py").write_text("# helper\n")
+        files = list(iter_scan_files([root]))
+        names = {p.name for p in files}
+        assert "test_pr_phase_alignment.py" in names
+        assert "_violation.py" in names

